File: crates/ref-id/examples/parse_lines.rs
// SPDX-License-Identifier: Apache-2.0
//
// Reads one identifier per line on stdin (with `\n` and `\r` escaped as `\\n` / `\\r`) and prints one
// canonical JSON result per line — the surface the differential test against the TypeScript reference reads.
//
// `--pairs` (Plan-005 Track 5) switches to a second protocol on the same escaping: stdin carries two
// lines per pair, `a` then `b`; stdout carries one line per pair, the canonical JSON of
// `{covers, coversReversed, samePackage, sameIdentifier, relate, verdict}`, where `coversReversed` is
// `covers(b, a)` and `relate`/`verdict` are the full result or `null`. Two lines per pair rather than a
// separator, because the grammar admits a tab in a locator.

use ref_id::{canonicalise, covers, parse, relate, same_identifier, same_package, serialise, verdict};
use serde_json::Value;
use std::io::BufRead;

/// `\n` and `\r` escaped as `\\n` / `\\r` — the one escaping both protocols share.
fn unescape(line: &str) -> String {
    line.replace("\\n", "\n").replace("\\r", "\r")
}

fn run_default(with_canonical: bool) -> i32 {
    let mut failures = 0;
    for line in std::io::stdin().lock().lines() {
        let input = unescape(&line.unwrap());
        match parse(&input) {
            Ok(result) => {
                let mut json = result.to_json();
                if let Some(map) = json.as_object_mut() {
                    if !with_canonical {
                        map.remove("canonical");
                    }
                    if let Ok(serialised) = serialise(&result) {
                        map.insert("serialised".into(), Value::String(serialised));
                    }
                }
                println!("{}", canonicalise(&json).unwrap());
            }
            Err(error) => {
                failures += 1;
                println!("{}", canonicalise(&serde_json::json!({ "threw": error.to_string() })).unwrap());
            }
        }
    }
    if failures == 0 {
        0
    } else {
        1
    }
}

/// Two lines in, one canonical JSON line out — `{covers, coversReversed, samePackage, sameIdentifier,
/// relate, verdict}` — for every complete pair. None of the six ever raises: `covers`/`samePackage`/
/// `sameIdentifier` answer `false` and `relate`/`verdict` answer `null` for whatever they refuse, so
/// there is nothing here to report as a per-pair failure. A trailing unpaired line — stdin ending after
/// an odd number of lines — has nothing to compare it against and is dropped rather than guessed at.
fn run_pairs() -> i32 {
    let mut lines = std::io::stdin().lock().lines();
    loop {
        let Some(a) = lines.next() else { break };
        let a = unescape(&a.unwrap());
        let Some(b) = lines.next() else { break };
        let b = unescape(&b.unwrap());

        let result = serde_json::json!({
            "covers": covers(&a, &b),
            "coversReversed": covers(&b, &a),
            "samePackage": same_package(&a, &b),
            "sameIdentifier": same_identifier(&a, &b),
            "relate": relate(&a, &b).map(|r| r.to_json()).unwrap_or(Value::Null),
            "verdict": verdict(&a, &b).map(|v| v.to_json()).unwrap_or(Value::Null),
        });
        println!("{}", canonicalise(&result).unwrap());
    }
    0
}

fn main() {
    // `--canonical` keeps the field the default protocol drops. The drop is deliberate: a locator's
    // validity belongs to the format, and the three purl libraries disagree at the edge, so the shared
    // protocol compares every field except that verdict. Canonicalisation is a different question —
    // two systems that compare identifiers by canonical form must agree on it — so it is measurable
    // here rather than silently excluded with the verdict.
    let args: Vec<String> = std::env::args().collect();
    let exit_code = if args.iter().any(|a| a == "--pairs") {
        run_pairs()
    } else {
        run_default(args.iter().any(|a| a == "--canonical"))
    };
    std::process::exit(exit_code);
}

File: packages/ref-id/parse-lines.ts
// SPDX-License-Identifier: Apache-2.0
//
// Reads one identifier per line on stdin (with `\n` and `\r` escaped as `\\n` / `\\r`) and prints one
// canonical JSON result per line — the same line protocol the Rust and Swift ports speak.
//
// It exists for parity, and parity here is not cosmetic. Without it the differential test called this
// package in-process while calling the other two as child processes, so the reference was judged by a
// path the ports never take: a defect living only in the protocol's own serialisation would have been
// invisible in exactly the implementation the others are compared against.
//
// `--canonical` keeps the delegated locator's canonical spelling. The default drops it, because the
// validity verdict of a Package URL belongs to the format and the three validators differ at the edge;
// canonicalisation is a separate question, and the two are not excluded together.
//
// `--browser` selects the browser entry point instead of the Node one — the same runner, the same
// protocol, one import apart (Plan-004, Track 3). That single difference is the point: the browser
// build gets its spec from a compiled-in constant and its sha256 from a different implementation, and
// the differential harness is what holds the two answers against each other. A suite of its own would
// compare each build against its own expectations, which is exactly the check that stayed green while
// the three language ports canonicalised a Package URL three different ways.
//
// `--pairs` (Plan-005, Track 5) switches to the second pass an adversarial review demanded: the four
// implementations agreed on every parse and still disagreed on comparisons no vector named. stdin then
// carries two lines per pair — `a`, then `b`, escaped the same way `--parse` escapes a line — and stdout
// carries one canonical-JSON line per pair: `covers`, `coversReversed` (`covers(b, a)`), `samePackage`,
// `sameIdentifier`, `relate` (the full result, or `null`) and `verdict` (the same). Two lines rather than one line with a
// separator, because the grammar admits a tab inside a locator.

const withCanonical = process.argv.includes("--canonical")
const asBrowser = process.argv.includes("--browser")
const pairsMode = process.argv.includes("--pairs")

// Dynamic, because the two entry points are the thing under test and a static import would pull both
// module graphs into this process — including the Node one, whose filesystem loader is precisely what
// the browser build does not have.
const { canonicalise, covers, parse, relate, samePackage, sameIdentifier, serialise, verdict } = asBrowser
  ? await import("./src/index.browser.ts")
  : await import("./src/index.ts")

const stdin: string = await new Promise((resolve) => {
  let text = ""
  process.stdin.setEncoding("utf8")
  process.stdin.on("data", (chunk) => (text += chunk))
  process.stdin.on("end", () => resolve(text))
})

const unescape = (line: string): string => line.replace(/\\n/g, "\n").replace(/\\r/g, "\r")

let failures = 0
const lines = stdin.split("\n").filter((line) => line.length > 0)

// Neither mode calls process.exit() after writing: stdout to a pipe is asynchronous, and exiting drops what is
// still queued. The pair pass lost 12 049 of 39 601 rows that way on the macOS runner the day the corpus grew
// by two vectors — the differential then reported the reference as not having run. `exitCode` lets the
// queue drain.
if (pairsMode && lines.length % 2 !== 0) {
  console.error(`parse-lines --pairs: ${lines.length} lines is not an even number of lines (two per pair)`)
  process.exitCode = 1
} else if (pairsMode) {
  for (let index = 0; index < lines.length; index += 2) {
    const a = unescape(lines[index]!)
    const b = unescape(lines[index + 1]!)
    console.log(
      canonicalise({
        covers: covers(a, b),
        coversReversed: covers(b, a),
        samePackage: samePackage(a, b),
        sameIdentifier: sameIdentifier(a, b),
        relate: relate(a, b),
        verdict: verdict(a, b),
      }),
    )
  }
} else {
  runParse()
}

function runParse(): void {
  for (const line of lines) {
    const input = unescape(line)
    let result
    try {
      result = parse(input)
    } catch (error) {
      failures += 1
      console.log(canonicalise({ threw: String(error) }))
      continue
    }
    const row: Record<string, unknown> = { ...result }
    // Shapes each implementation forms for itself, and never part of the comparison.
    delete row.nested
    delete row.delegated
    if (!withCanonical) {
      delete row.canonical
    }
    try {
      row.serialised = serialise(result)
    } catch {
      // A result the serialiser refuses carries no `serialised` field — the ports do the same.
    }
    console.log(canonicalise(row))
  }
  process.exitCode = failures === 0 ? 0 : 1
}
